### PR TITLE
Authorization with nonce

### PR DESCRIPTION
**Pull request by cgilbert76**
Move to new authorization flow, where a nonce is passed to the device platform rather than sensitive data such as an external patient ID or system identifier.  A nonce is generated and used to create a blob store, which is then used to contain the authorization data.  Upon callback from the device platform, the passed back nonce is used to retrieve the authorization data and complete the process.

### CODE REVIEW

File: src/Common/FitOnFhir.Common/Microsoft.Health.FitOnFhir.Common.csproj [hunk 1]
```diff
@@ -13,6 +13,7 @@
   <ItemGroup>
     <PackageReference Include="Azure.Data.Tables" Version="12.5.0" />
     <PackageReference Include="Azure.Security.KeyVault.Secrets" Version="4.2.0" />
+    <PackageReference Include="Azure.Storage.Blobs" Version="12.13.0" />
     <PackageReference Include="Azure.Storage.Queues" Version="12.11.0" />
```
Comment (namalu) [inline src/Common/FitOnFhir.Common/Microsoft.Health.FitOnFhir.Common.csproj:17]: Is there a new version of Azure.Storage.Queues that matches the Azure.Storage.Blobs version number 12.13.0? If so we should update it.

Comment (cgilbert76) [reply src/Common/FitOnFhir.Common/Microsoft.Health.FitOnFhir.Common.csproj:17]: There is not.  12.11.0 is the latest for Azure.Storage.Queues

File: src/Common/FitOnFhir.Common/Services/AuthStateService.cs [hunk 2]
```diff
@@ -13,20 +15,37 @@
...
         public AuthStateService(
+            AzureConfiguration azureConfiguration,
             AuthenticationConfiguration authenticationConfiguration,
             IJwtSecurityTokenHandlerProvider jwtSecurityTokenHandlerProvider,
+            BlobServiceClient blobServiceClient,
             ILogger<AuthStateService> logger)
         {
+            if (blobServiceClient == null)
+            {
+                _blobServiceClient = new BlobServiceClient(azureConfiguration.StorageAccountConnectionString);
```
Comment (namalu) [inline src/Common/FitOnFhir.Common/Services/AuthStateService.cs:41]: nit: We probably want to check if the azureConfiguration.StorageAccountConnectionString is null,

File: src/Common/FitOnFhir.Common/Services/AuthStateService.cs [hunk 3]
```diff
@@ -81,5 +100,62 @@ public AuthState CreateAuthState(HttpRequest httpRequest)
...
+        {
+            try
+            {
+                var nonce = GenerateNonce(Constants.NonceLength);
+
+                // Get a reference to the blob
+                BlobClient blobClient = _blobContainerClient.GetBlobClient(nonce);
+
+                // Set the CacheControl property to expire in 5 minutes (300 seconds)
+                await blobClient.SetHttpHeadersAsync(new BlobHttpHeaders { CacheControl = "max-age=300" }, cancellationToken: cancellationToken);
```
Comment (namalu) [inline src/Common/FitOnFhir.Common/Services/AuthStateService.cs:136]: I don't think the CacheControl header is what you want to use. I believe this is a way for clients to expire local caches. This won't actually delete the blob after 5 minutes.

I think this is what you want:

https://docs.microsoft.com/en-us/rest/api/storageservices/set-blob-expiry#expiryoption

Comment (cgilbert76) [reply src/Common/FitOnFhir.Common/Services/AuthStateService.cs:136]: This is what I had found, that led me to put that in:
https://docs.microsoft.com/en-us/azure/cdn/cdn-manage-expiration-of-blob-content?toc=%2Fazure%2Ffrontdoor%2FTOC.json

Comment (namalu) [reply src/Common/FitOnFhir.Common/Services/AuthStateService.cs:136]: That document is a very specific use case when a Content Distribution Network (CDN) is connected to Azure Blob Store. This provides a way for the CDN to cache the blob internally and serve it to consumers for a specified period of time without making a request to blob every time. Once the max-age is reached, the CDN will fetch the blob again and cache that version of the blob until the max-age expires the blob again.

In other words, the blob is never deleted, it just provides a way to control how often a consumer needs to update their cache.

File: src/Common/FitOnFhir.Common/Interfaces/IAuthStateService.cs [hunk 4]
```diff
@@ -22,5 +22,22 @@ public interface IAuthStateService
         /// Will throw an <see cref="ArgumentException"/> if <see cref="AuthenticationConfiguration.IsAnonymousLoginEnabled"/> is false and
         /// query params <see cref="Constants.ExternalIdQueryParameter"/> and <see cref="Constants.ExternalSystemQueryParameter"/> are present in the <see cref="HttpRequest"/>.</returns>
         AuthState CreateAuthState(HttpRequest httpRequest);
```
Comment (namalu) [inline src/Common/FitOnFhir.Common/Interfaces/IAuthStateService.cs:24]: I am seeing that the `CreateAuthState()` method seems more like an internal member. Since the AuthState object is only used externally for the Callback request, It might make sense to encapsulate the creation of a new object in the AuthStateService.

Comment (namalu) [reply src/Common/FitOnFhir.Common/Interfaces/IAuthStateService.cs:24]: Never mind. I see now that you are using this for the `Revoke()` flow too.

File: src/Common/FitOnFhir.Common/Services/AuthStateService.cs [hunk 5]
```diff
@@ -81,5 +100,62 @@ public AuthState CreateAuthState(HttpRequest httpRequest)
 
             return new AuthState(ExternalIdentifier, ExternalSystem);
         }
+
+        public async Task<AuthState> RetrieveAuthState(string nonce, CancellationToken cancellationToken)
+        {
+            try
+            {
+                // Get a reference to the blob
+                BlobClient blobClient = _blobContainerClient.GetBlobClient(nonce);
```
Comment (namalu) [inline src/Common/FitOnFhir.Common/Services/AuthStateService.cs:109]: I think we need to make sure that this is not null - OR - does this throw if `GetBlobClient()` is called with a nonce that does not exist?

File: src/GoogleFit/FitOnFhir.GoogleFit/Services/UsersService.cs [hunk 6]
```diff
@@ -48,32 +49,24 @@ public class UsersService : UsersServiceBase, IUsersService
...
-            {
-                authState = AuthState.Parse(state);
-            }
-            catch (Exception e) when (e is ArgumentException || e is JsonSerializationException)
-            {
-                // This exception message will be visible to the caller.
-                throw new Exception("The Google authorization service failed to return the expected authorization state.");
-            }
+            var validNonce = EnsureArg.IsNotNullOrWhiteSpace(nonce);
+            AuthState authState = await _authStateService.RetrieveAuthState(validNonce, cancellationToken);
```
Comment (namalu) [inline src/GoogleFit/FitOnFhir.GoogleFit/Services/UsersService.cs:69]: I think we need to catch any Exception that might happen when we try to retrieve the AuthState using the nonce and throw an Exception with an error that is appropriate for the general public. You can see in the error you deleted above, we are trying to make sure that public facing errors are more polished.

File: src/Common/FitOnFhir.Common/Services/AuthStateService.cs (hunk 3 above)
Comment (cgilbert76) [reply src/Common/FitOnFhir.Common/Services/AuthStateService.cs:136]: From the docs link you shared, note that it only applies to Hierarchical Namespace enabled accounts.  A quick check of our storage account resource created by the template reveals that Hierarchical Namespaces are disabled.  If I attempt to enable that, I am redirected to a page which talks about upgrading "to a storage account with Azure Data Lake Gen2 capabilities".

This is not looking very cut and dried.  I am also surprised and disappointed, that there is no direct C# method or property to do something as simple as set an expiration for a blob.

File: src/GoogleFit/FitOnFhir.GoogleFit/Services/UsersService.cs (hunk 6 above)
Comment (cgilbert76) [reply src/GoogleFit/FitOnFhir.GoogleFit/Services/UsersService.cs:69]: Yes, in `RetrieveAuthState` I catch any exception and throw.  This is ultimately caught again in `GoogleFitAuthorizationHandler.Callback`, which returns a 404.  

I deleted the exception shown here, because that message is no longer accurate 100% of the time.  Now there could be a failure either on Google's end with the returned nonce, or in the retrieval of the blob.  

I am open to suggestions on what the exception message should be.  I kept it general for now, because I didn't know of a better way to phrase it ("Failed to retrieve the auth state.")

**Comment by cgilbert76** [comment]
The pipeline for this pull request keeps failing.  The error occurs during the Deploying Ephemeral stage, with the message 'The template parameters 'spid' in the parameters file are not valid; they are not present in the original template and can therefore not be provided at deployment time.'

After doing much digging, I believe it is referring to a line in the deploy-parse-arm.yml file where there appears to be a command line switch 'spid':
`overrideParameters: >
        ${{ parameters.overrideParameters }}
        -spid $(getObjectId.objectId)`

Unable to find any sort of documentation on what that line is doing, I resorted to slowly backing out anything new that I added to azuredeploy.bicep.  It still failed after all that was left is the `authentication_blob_container_name` declaration.  I highly doubt that alone is the cause of the failure.  I am declaring it and using it just like any other param in the .bicep file.

It would appear that whatever this vague error is, is likely the root cause for why I can't deploy from a template to my resource group.  The same three resources keep failing (`import-data-foftst/web`, `authorize-foftst/web`, and `import-timer-foftst/web`.  I say that it is likely, because that foftst basename is completely new, so there should not be any resource conflicts with older resources that haven't been deleted or purged from previous deployments.

File: src/Common/FitOnFhir.Common/Services/AuthStateService.cs (hunk 3 above)
Comment (namalu) [reply src/Common/FitOnFhir.Common/Services/AuthStateService.cs]: What happens if you use the storage account as is and just set the `x-ms-expiry-option` header?

File: src/GoogleFit/FitOnFhir.GoogleFit/Services/UsersService.cs (hunk 6 above)
Comment (namalu) [reply src/GoogleFit/FitOnFhir.GoogleFit/Services/UsersService.cs:69]: I think it would be fine to just say "An error occurred while validating the Google authorization callback 'state' parameter."

**Review by namalu — APPROVED** [review]
